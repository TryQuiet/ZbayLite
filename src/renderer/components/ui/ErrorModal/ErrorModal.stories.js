import React from 'react'
import { storiesOf } from '@storybook/react'
import { withKnobs, text } from '@storybook/addon-knobs'
import { action } from '@storybook/addon-actions'

import ErrorModal from './ErrorModal'

storiesOf('Components/UI/ErrorModal', module)
  .addDecorator(withKnobs)
  .add('playground', () => {
    try {
      throw new Error('Something went really badly')
    } catch (error) {
      return (
        <ErrorModal
          open
          message={text('Error message', error.message)}
          traceback={text('Traceback', error.stack)}
          handleClose={action('Handle close')}
          handleCopy={action('Handle copy')}
        />
      )
    }
  })
