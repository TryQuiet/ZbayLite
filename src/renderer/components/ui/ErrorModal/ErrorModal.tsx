import React from 'react'

import fetch from 'isomorphic-fetch'

import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import TextField from '@material-ui/core/TextField'

import { makeStyles } from '@material-ui/core/styles'

import red from '@material-ui/core/colors/red'

import Icon from '../Icon/Icon'
import Modal from '../Modal/Modal'
import LoadingButton from '../LoadingButton/LoadingButton'

import { LOG_ENDPOINT } from '../../../../shared/static'

import exclamationMark from '../../../static/images/exclamationMark.svg'

const useStyles = makeStyles(theme => ({
  root: {
    padding: theme.spacing(4)
  },
  icon: {
    fontSize: '10rem',
    color: red[500],
    width: 80,
    height: 70
  },
  stackTrace: {
    fontSize: '14px',
    wordBreak: 'break-all'
  },
  message: {
    wordBreak: 'break-all',
    marginTop: 20,
    fontWeight: 500
  },
  info: {
    textAlign: 'center'
  },
  textfield: {},
  cssDisabled: {
    backgroundColor: theme.palette.colors.inputGray,
    color: theme.palette.colors.red
  },
  button: {
    textTransform: 'none',
    width: 150,
    height: 60
  }
}))

const handleSend = async ({ title, message }: { title: string; message: string }) => {
  await fetch(LOG_ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      title,
      message
    })
  })
}

interface ErrorModalProps {
  open: boolean
  message: string
  traceback: string
  handleExit: () => void
  restartApp: () => void
}

export const ErrorModal: React.FC<ErrorModalProps> = ({
  open = false,
  message,
  traceback,
  handleExit,
  restartApp
}) => {
  const classes = useStyles({})

  const [send, setSend] = React.useState(false)

  return (
    <Modal open={open} handleClose={handleExit} title='Error'>
      <Grid container justify='flex-start' spacing={3} direction='column' className={classes.root}>
        <Grid item container direction='column' alignItems='center'>
          <Icon className={classes.icon} src={exclamationMark} />
          <Typography variant='h3' className={classes.message}>
            {message}
          </Typography>
        </Grid>
        <Grid item container spacing={2} direction='column'>
          <Grid item>
            <Typography variant='body2' className={classes.info}>
              You can send us this error traceback to help us improve. Before sending make sure it
              doesn't contain any private data.
            </Typography>
          </Grid>
          <Grid item>
            <TextField
              id='traceback'
              disabled
              variant='outlined'
              multiline
              fullWidth
              rows={10}
              value={traceback}
              InputProps={{
                classes: {
                  root: classes.textfield,
                  multiline: classes.stackTrace,
                  disabled: classes.cssDisabled
                }
              }}
            />
          </Grid>
          <Grid item container justify='center' alignItems='center'>
            {!send && (
              <LoadingButton
                text='Send & restart'
                classes={{ button: classes.button }}
                onClick={async () => {
                  try {
                    await handleSend({ title: message, message: traceback })
                    setSend(true)
                    restartApp()
                  } catch (err) {
                    console.log('ERROR SENDING MESSAGE', err.message)
                  }
                }}
              />
            )}
          </Grid>
        </Grid>
      </Grid>
    </Modal>
  )
}

export default ErrorModal
