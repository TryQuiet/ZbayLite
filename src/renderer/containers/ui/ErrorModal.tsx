import React from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import * as R from 'ramda'
import { remote } from 'electron'

import ErrorModal from '../../components/ui/ErrorModal'
import criticalErrorSelectors from '../../store/selectors/criticalError'
import modalsHandlers, { withModal } from '../../store/handlers/modals'
import notificationsHandlers from '../../store/handlers/notifications'
import {
  successNotification,
  errorNotification
} from '../../store/handlers/utils'

export const mapStateToProps = state => ({
  message: criticalErrorSelectors.message(state),
  traceback: criticalErrorSelectors.traceback(state)
})

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      handleExit: modalsHandlers.actionCreators.openModal('quitApp'),
      successSnackbar: () =>
        notificationsHandlers.actions.enqueueSnackbar(
          successNotification({
            message: 'Message has been sent'
          })
        ),
      errorSnackbar: () =>
        notificationsHandlers.actions.enqueueSnackbar(
          errorNotification({
            message: 'There was an error'
          })
        ),
      restartApp: () => {
        remote.app.relaunch()
        remote.app.quit()
      }
    },

    dispatch
  )

export default R.compose(
  React.memo,
  connect(mapStateToProps, mapDispatchToProps),
  withModal('criticalError')
)(ErrorModal)
