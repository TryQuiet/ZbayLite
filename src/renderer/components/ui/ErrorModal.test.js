/* eslint import/first: 0 */
import React from 'react'
import { shallow } from 'enzyme'

import { ErrorModal } from './ErrorModal'

describe('ErrorModal', () => {
  it('renders component', () => {
    const result = shallow(
      <ErrorModal
        open
        message='Test error message'
        traceback='Error: Test error message, error traceback'
        handleExit={jest.fn()}
        handleCopy={jest.fn()}
        restartApp={jest.fn()}
      />
    )
    expect(result).toMatchSnapshot()
  })
})
