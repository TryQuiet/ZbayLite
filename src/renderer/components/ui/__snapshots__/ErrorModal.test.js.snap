// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorModal renders component 1`] = `
<Modal
  addBorder={false}
  alignCloseLeft={false}
  canGoBack={false}
  contentWidth={600}
  handleClose={[MockFunction]}
  isBold={false}
  isCloseDisabled={false}
  open={true}
  title="Error"
>
  <WithStyles(ForwardRef(Grid))
    className="root-mock"
    container={true}
    direction="column"
    justify="flex-start"
    spacing={3}
  >
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      container={true}
      direction="column"
      item={true}
    >
      <Icon
        className="icon-mock"
        src="test-file-stub"
      />
      <WithStyles(ForwardRef(Typography))
        className="message-mock"
        variant="h3"
      >
        Test error message
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      container={true}
      direction="column"
      item={true}
      spacing={2}
    >
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(Typography))
          className="info-mock"
          variant="body2"
        >
          You can send us this error traceback to help us improve. Before sending make sure it doesn't contain any private data.
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        item={true}
      >
        <WithStyles(ForwardRef(TextField))
          InputProps={
            Object {
              "classes": Object {
                "disabled": "cssDisabled-mock",
                "multiline": "stackTrace-mock",
                "root": "textfield-mock",
              },
            }
          }
          disabled={true}
          fullWidth={true}
          id="traceback"
          multiline={true}
          rows={10}
          value="Error: Test error message, error traceback"
          variant="outlined"
        />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        alignItems="center"
        container={true}
        item={true}
        justify="center"
      >
        <WithStyles(LoadingButton)
          classes={
            Object {
              "button": "button-mock",
            }
          }
          onClick={[Function]}
          text="Send & restart"
        />
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</Modal>
`;
